File: vol1/ch7/LeetCode272-2.py
from collections import deque
class Solution:
    def closestKValues(self,root:Optional[TreeNode],target:float,k:int)->List[int]:
        self.qu=deque()             #作为队列
        self.inorder(root,target,k)
        return list(self.qu)

    def inorder(self,root,target,k):
        if root:
            self.inorder(root.left,target,k)
            if len(self.qu)==k:
                if abs(self.qu[0]-target)>abs(root.val-target):
                    self.qu.popleft()
                    self.qu.append(root.val)
                else: return
            else: self.qu.append(root.val)
            self.inorder(root.right,target,k)

File: vol1/ch7/LeetCode272-2.cpp
class Solution {
public:
    vector<int> closestKValues(TreeNode* root,double target,int k) {
        queue<int> qu;
        inorder(qu,root,target,k);
        vector<int> ans;
        while(!qu.empty()) {
            ans.push_back(qu.front());
            qu.pop();
        }
        return ans;
    }

    void inorder(queue<int>&qu,TreeNode*root,double target,int k) {
        if(root!=NULL) {
            inorder(qu,root->left,target,k);
            if(qu.size()==k) {
                if(abs(qu.front()-target)>abs(root->val-target)) {
                    qu.pop();
                    qu.push(root->val);
                }
                else return;
            }
            else qu.push(root->val);
            inorder(qu,root->right,target,k);
        }
    }
};

